File: Graph Functionalities/ui/ui.py


class Ui:
    def __init__(self, service):
        self._service = service

    def _print_menu(self):
        print("Menu")
        print("1. Generate random graph")
        print("2. Read from file")
        print("3. Write to file directed")
        print("4. Print directed graph")
        print("5. Add vertex")
        print("6. Remove vertex")
        print("7. Add edge directed")
        print("8. Remove edge directed")
        print("9. In degree")
        print("10. Out degree")
        print("11. Outbound vertices")
        print("12. Inbound vertices")
        print("13. Parse vertices")
        print("14. Modify information")
        print('15. Vertex number')
        print('16. Edge number')
        print('17. Copy graph')
        print('18. Print graph as tree')
        print('19. Change to copy')
        print('20. Dijkstra')
        print('21. Add undirected edge')
        print('22. Remove undirected edge')
        print('23. Read undirected from file')
        print('24. Print undirected graph')
        print('25. BFS connected')
        print('26. Generate random undirected graph')
        print('27. Write to file undirected')
        print('28. Lowest cost walk between 2 vertices (Floyd Warshall)')
        print('29. Maximum clique')
        print("0. Exit")

    def _in_degree(self):
        x = input("\nEnter the vertex number: ")

        print(self._service.in_degree(x))

    def _out_degree(self):
        x = input("\nEnter the vertex number: ")

        print(self._service.out_degree(x))

    def _random_ui(self):
        n = input("\nEnter the number of vertices: ")
        m = input("Enter the number of edges: ")

        self._service.generate_random(n, m)

    def _read_from_file(self):
        file_name = input("\nEnter the file name: ")

        self._service.read_from_file(file_name)

    def _write_to_file(self):
        file_name = input("\nEnter the file name: ")

        self._service.write_to_file(file_name)

    def _print_graph(self):
        print(self._service.readable_graph())

    def _add_vertex_ui(self):
        x = input("\nEnter the vertex number: ")

        self._service.add_vertex(x)

    def _remove_vertex_ui(self):
        x = input("\nEnter the vertex number: ")

        self._service.remove_vertex(x)

    def _add_edge_ui(self):
        x = input("\nEnter the source vertex: ")
        y = input("Enter the destination vertex: ")
        val = input("Enter the information on the edge: ")

        self._service.add_edge(x, y, val)

    def _remove_edge_ui(self):
        x = input("\nEnter the source vertex: ")
        y = input("Enter the destination vertex: ")

        self._service.remove_edge(x, y)

    def _outbound(self):
        x = input("\nEnter the source vertex: ")

        string = str(x) + ': '
        for y in self._service.parse_out(x):
            string += str(y) + ' '
        print(string)

    def _inbound(self):
        x = input("\nEnter the destination vertex: ")

        string = str(x) + ': '
        for y in self._service.parse_in(x):
            string += str(y) + ' '
        print(string)

    def _print_vertices(self):
        string = ''
        for x in self._service.parse_all():
            string += str(x) + ' '
        print(string)

    def _modify(self):
        x = input("\nEnter the source vertex: ")
        y = input("Enter the destination vertex: ")
        val = input("Enter the information on the edge: ")

        self._service.modify_info(x, y, val)

    def _vertex_number(self):
        print(self._service.vertices_number())

    def _edge_number(self):
        print(self._service.edges_number())

    def _copy(self):
        self._service.copy_graph()

    def _print_tree(self):
        print(self._service.get_graph())

    def _change_copy(self):
        self._service.change_to_copy()

    def _dijkstra(self):
        x = int(input("Enter the beginning vertex: "))
        prev = self._service.dijkstra(x)

    def _addU(self):
        x = input("\nEnter the source vertex: ")
        y = input("Enter the destination vertex: ")
        val = input("Enter the information on the edge: ")

        self._service.add_edge_undirected(x, y)

    def _removeU(self):
        x = input("\nEnter the source vertex: ")
        y = input("Enter the destination vertex: ")

        self._service.remove_edge_undirected(x, y)

    def _readU(self):
        file_name = input("\nEnter the file name: ")

        self._service.read_from_file_undirected(file_name)

    def _printU(self):
        print(self._service.readable_undirected())

    def _BFSconnected(self):
        cc = self._service.connected()
        for g in cc:
            print(g.parseX())
            print(g.print_undirected(), '\n')

    def _random_undir(self):
        n = input("\nEnter the number of vertices: ")
        m = input("Enter the number of edges: ")

        self._service.generate_random_undirected(n, m)

    def _write_undirected(self):
        fn = input("Enter file name: ")
        self._service.write_to_file_undirected(fn)

    def _Floyd_Warshall(self):
        x = input("\nEnter source vertex: ")
        y = input("Enter destination vertex: ")
        path, cost = self._service.Floyd_Warshall(x, y)
        print('cost = ' + str(cost))
        print('path = ' + ' -> '.join(str(i) for i in path))

    def _max_clique(self):
        print(" ".join(str(v) for v in self._service.max_clique()))

    def start(self):
        is_running = True

        command_dict = {'1': self._random_ui, '2': self._read_from_file, '3': self._write_to_file,
                        '4': self._print_graph, '5': self._add_vertex_ui, '6': self._remove_vertex_ui,
                        '7': self._add_edge_ui, '8': self._remove_edge_ui, '9': self._in_degree, '10': self._out_degree,
                        '11': self._outbound, '12': self._inbound, '13': self._print_vertices, '14': self._modify,
                        '15': self._vertex_number, '16': self._edge_number, '17': self._copy, '18': self._print_tree,
                        '19': self._change_copy, '20': self._dijkstra, '21': self._addU, '22': self._removeU,
                        '23': self._readU, '24': self._printU, '25': self._BFSconnected, '26': self._random_undir,
                        '27': self._write_undirected, '28': self._Floyd_Warshall, '29': self._max_clique}

        while is_running:
            self._print_menu()

            command = input(">> ")
            try:
                if command in command_dict:
                    command_dict[command]()
                elif command == '0':
                    is_running = False
                else:
                    print("\nInvalid command\n")
            except Exception as ex:
                print(str(ex))

File: Graph Functionalities/Graph practical1/Graph practical1/ui.cpp
#include "ui.h"
#include <iostream>

void Ui::print_menu()
{
	cout << "Menu\n";
	cout << "1. Generate random graph\n";
	cout << "2. Read from file\n";
	cout << "3. Write to file\n";
	cout << "4. Print graph\n";
	cout << "5. Add vertex\n";
	cout << "6. Remove vertex\n";
	cout << "7. Add edge\n";
	cout << "8. Remove edge\n";
	cout << "9. Outbound vertices\n";
	cout << "10. Inbound vertices\n";
	cout << "11. Parse vertices\n";
	cout << "12. Modify information\n";
	cout << "13. Vertex number\n";
	cout << "14. Edge number\n";
	cout << "15. Copy graph\n";
	cout << "16. Change to copy graph\n";
	cout << ">> ";
}

void Ui::random_ui()
{
	cin.get();
	char n[255], m[255];

	cout << "\nEnter the number of vertices: ";
	cin.getline(n, 255);

	cout << "\nEnter the number of edges: ";
	cin.getline(m, 255);

	string n_s(n);
	string m_s(m);

	this->service.generate_random(n_s, m_s);
}

void Ui::read_from_file_ui()
{
	cin.get();
	char file_name[255];

	cout << "\nEnter file name: ";
	cin.getline(file_name, 255);

	string fn(file_name);

	this->service.read_from_file(fn);
}

void Ui::write_to_file_ui()
{
	cin.get();
	char file_name[255];

	cout << "\nEnter file name: ";
	cin.getline(file_name, 255);

	string fn(file_name);

	this->service.write_to_file(fn);
}

void Ui::print_graph()
{
	cout << this->service.write_graph();
}

void Ui::add_vertex_ui()
{
	cin.get();
	char x[255];

	cout << "\nEnter the vertex: ";
	cin.getline(x, 255);

	string x_s(x);

	int r = this->service.add_vertex(x_s);

	if (r == 1)
		cout << "\nVertex already exists\n";
}

void Ui::remove_vertex_ui()
{
	cin.get();
	char x[255];

	cout << "\nEnter the vertex: ";
	cin.getline(x, 255);

	string x_s(x);

	int r = this->service.remove_vertex(x_s);

	if (r == 1)
		cout << "\nVertex not found\n";
}

void Ui::add_edge_ui()
{
	cin.get();
	char x[255], y[255], val[255];

	cout << "\nEnter the source vertex: ";
	cin.getline(x, 255);

	cout << "\nEnter the destination vertex: ";
	cin.getline(y, 255);

	cout << "\nEnter the value: ";
	cin.getline(val, 255);

	string x_s(x);
	string y_s(y);
	string val_s(val);

	int r = this->service.add_edge(x_s, y_s, val_s);

	if (r == 1)
		cout << "\nEdge already exists or vertices not found\n";
}

void Ui::remove_edge_ui()
{
	cin.get();
	char x[255], y[255];

	cout << "\nEnter the source vertex: ";
	cin.getline(x, 255);

	cout << "\nEnter the destination vertex: ";
	cin.getline(y, 255);

	string x_s(x);
	string y_s(y);

	int r = this->service.remove_edge(x_s, y_s);

	if (r == 1)
		cout << "\nEdge does not exist\n";
}

void Ui::outbound()
{
	cin.get();
	char x[255];

	cout << "\nEnter the source vertex: ";
	cin.getline(x, 255);

	string x_s(x);

	vector<int> ver = this->service.parse_out(x_s);

	string str = x_s + ": ";

	for (int i = 0; i < ver.size(); i++)
		str += " " + to_string(ver[i]);

	cout << str << '\n';
}

void Ui::inbound()
{
	cin.get();
	char x[255];

	cout << "\nEnter the destination vertex: ";
	cin.getline(x, 255);

	string x_s(x);

	vector<int> ver = this->service.parse_in(x_s);

	string str = x_s + ": ";

	for (int i = 0; i < ver.size(); i++)
		str += " " + to_string(ver[i]);

	cout << str << '\n';
}

void Ui::print_vertices()
{
	cin.get();

	vector<int> ver = this->service.parse_all();

	for (int i = 0; i < ver.size(); i++)
		cout << ver[i] << ' ';

	cout << '\n';
}

void Ui::modify()
{
	cin.get();
	char x[255], y[255], val[255];

	cout << "\nEnter the source vertex: ";
	cin.getline(x, 255);

	cout << "\nEnter the destination vertex: ";
	cin.getline(y, 255);

	cout << "\nEnter the new value: ";
	cin.getline(val, 255);

	string x_s(x);
	string y_s(y);
	string val_s(val);

	int r = this->service.modify_info(x_s, y_s, val_s);

	if (r == 1)
		cout << "\nEdge does not exist\n";
}

void Ui::vertex_number()
{
	cout << this->service.vertices_number();
}

void Ui::edge_number()
{
	cout << this->service.edges_number();
}

void Ui::copy()
{
	this->service.copy_graph();
}

void Ui::change_to_copy()
{
	this->service.change_to_copy();
}

Ui::Ui(Service& service) : service(service)
{
}

void Ui::start()
{
	bool is_runing = true;
	int command;

	while (is_runing)
	{
		this->print_menu();

		cin >> command;

		switch (command)
		{
		case 0:
			is_runing = false;
			break;
		case 1:
			this->random_ui();
			break;

		case 2:
			this->read_from_file_ui();
			break;

		case 3:
			this->write_to_file_ui();
			break;

		case 4:
			this->print_graph();
			break;

		case 5:
			this->add_vertex_ui();
			break;

		case 6:
			this->remove_vertex_ui();
			break;

		case 7:
			this->add_edge_ui();
			break;

		case 8:
			this->remove_edge_ui();
			break;

		case 9:
			this->outbound();
			break;

		case 10:
			this->inbound();
			break;

		case 11:
			this->print_vertices();
			break;

		case 12:
			this->modify();
			break;

		case 13:
			this->vertex_number();
			break;

		case 14:
			this->edge_number();
			break;

		case 15:
			this->copy();
			break;

		case 16:
			this->change_to_copy();
			break;

		default:
			cout << "\nInvalid command\n";
			break;
		}
	}
}
